File: C++/structs/huffman/huffman.cpp
#include "huffman.h"
#include <iostream>
#include <sstream>

using namespace std;

HuffmanTree::HuffmanTree(const string& text) : root(nullptr) {
    if (text.empty()) {
        return;
    }
    
    // Подсчет частот символов
    map<char, int> frequencies;
    for (char c : text) {
        frequencies[c]++;
    }
    
    // Создание очереди с приоритетом
    priority_queue<HuffmanNode*, vector<HuffmanNode*>, CompareNode> pq;
    
    for (auto& pair : frequencies) {
        pq.push(new HuffmanNode(pair.first, pair.second));
    }
    
    // Построение дерева Хаффмана
    while (pq.size() > 1) {
        HuffmanNode* left = pq.top();
        pq.pop();
        HuffmanNode* right = pq.top();
        pq.pop();
        
        HuffmanNode* parent = new HuffmanNode('\0', left->frequency + right->frequency);
        parent->left = left;
        parent->right = right;
        
        pq.push(parent);
    }
    
    root = pq.top();
    
    // Генерация кодов
    generateCodes(root, "");
}

HuffmanTree::~HuffmanTree() {
    deleteTree(root);
}

void HuffmanTree::deleteTree(HuffmanNode* node) {
    if (node == nullptr) {
        return;
    }
    deleteTree(node->left);
    deleteTree(node->right);
    delete node;
}

void HuffmanTree::generateCodes(HuffmanNode* node, const string& code) {
    if (node == nullptr) {
        return;
    }
    
    // Если это лист, сохраняем код
    if (node->left == nullptr && node->right == nullptr) {
        codes[node->ch] = code;
        return;
    }
    
    // Рекурсивно обходим дерево
    generateCodes(node->left, code + "0");
    generateCodes(node->right, code + "1");
}

string HuffmanTree::encode(const string& text) const {
    if (root == nullptr) {
        return "";
    }
    
    ostringstream encoded;
    for (char c : text) {
        auto it = codes.find(c);
        if (it != codes.end()) {
            encoded << it->second;
        }
    }
    return encoded.str();
}

string HuffmanTree::decode(const string& encoded) const {
    if (root == nullptr || encoded.empty()) {
        return "";
    }
    
    ostringstream decoded;
    HuffmanNode* current = root;
    
    for (char bit : encoded) {
        if (bit == '0') {
            current = current->left;
        } else if (bit == '1') {
            current = current->right;
        }
        
        if (current->left == nullptr && current->right == nullptr) {
            decoded << current->ch;
            current = root;
        }
    }
    
    return decoded.str();
}

void HuffmanTree::printCodes() const {
    if (codes.empty()) {
        cout << "Нет кодов для отображения" << endl;
        return;
    }
    
    cout << "Коды символов Хаффмана:" << endl;
    for (const auto& pair : codes) {
        cout << pair.first << ": " << pair.second << endl;
    }
}

void HuffmanTree::printTree() const {
    cout << "Дерево Хаффмана:" << endl;
    printNode(root, "");
}

void HuffmanTree::printNode(HuffmanNode* node, const string& prefix) const {
    if (node == nullptr) {
        return;
    }
    
    if (node->left == nullptr && node->right == nullptr) {
        cout << prefix << "'" << node->ch << "' (freq: " << node->frequency << ")" << endl;
        return;
    }
    
    cout << prefix << "* (freq: " << node->frequency << ")" << endl;
    
    if (node->left != nullptr) {
        printNode(node->left, prefix + "  ");
    }
    if (node->right != nullptr) {
        printNode(node->right, prefix + "  ");
    }
}

double HuffmanTree::getCompressionRatio(const string& originalText) const {
    if (originalText.empty()) {
        return 0;
    }
    
    int originalBits = originalText.length() * 8;
    int encodedBits = encode(originalText).length();
    
    if (encodedBits == 0) {
        return 0;
    }
    
    return static_cast<double>(originalBits) / encodedBits;
}

File: go/structs/huffman/huffman.go
package huffman

import (
	"container/heap"
	"fmt"
	"strings"
)

type Node struct {
	Char      rune
	Frequency int
	Left      *Node
	Right     *Node
}

type PriorityQueue []*Node

// реализуем интерфейс для heap
func (pq PriorityQueue) Len() int {
	return len(pq)
}

func (pq PriorityQueue) Less(i, j int) bool {
	return pq[i].Frequency < pq[j].Frequency
}

func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
}

func (pq *PriorityQueue) Push(x interface{}) {
	*pq = append(*pq, x.(*Node))
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	node := old[n-1]
	*pq = old[0 : n-1]
	return node
}

type HuffmanTree struct {
	Root  *Node
	Codes map[rune]string
}

func NewHuffmanTree(text string) *HuffmanTree {
	if text == "" {
		return &HuffmanTree{
			Root:  nil,
			Codes: make(map[rune]string),
		}
	}

	frequencies := make(map[rune]int)
	for _, char := range text {
		frequencies[char]++
	}

	pq := make(PriorityQueue, 0, len(frequencies))
	for char, freq := range frequencies {
		pq = append(pq, &Node{
			Char:      char,
			Frequency: freq,
			Left:      nil,
			Right:     nil,
		})
	}

	heap.Init(&pq)

	// строим дерево
	for pq.Len() > 1 {
		left := heap.Pop(&pq).(*Node)
		right := heap.Pop(&pq).(*Node)

		parent := &Node{
			Char:      0,
			Frequency: left.Frequency + right.Frequency,
			Left:      left,
			Right:     right,
		}

		heap.Push(&pq, parent)
	}

	root := heap.Pop(&pq).(*Node)

	tree := &HuffmanTree{
		Root:  root,
		Codes: make(map[rune]string),
	}

	tree.generateCodes(root, "")

	return tree
}

func (ht *HuffmanTree) generateCodes(node *Node, code string) {
	if node == nil {
		return
	}

	// Если это лист, сохраняем код
	if node.Left == nil && node.Right == nil {
		ht.Codes[node.Char] = code
		return
	}

	// Рекурсивно обходим дерево
	ht.generateCodes(node.Left, code+"0")
	ht.generateCodes(node.Right, code+"1")
}

func (ht *HuffmanTree) Encode(text string) string {
	if ht.Root == nil {
		return ""
	}

	var encoded strings.Builder
	for _, char := range text {
		if code, exists := ht.Codes[char]; exists {
			encoded.WriteString(code)
		}
	}
	return encoded.String()
}

func (ht *HuffmanTree) Decode(encoded string) string {
	if ht.Root == nil || encoded == "" {
		return ""
	}

	var decoded strings.Builder
	current := ht.Root

	for _, bit := range encoded {
		if bit == '0' {
			current = current.Left
		} else if bit == '1' {
			current = current.Right
		}

		if current.Left == nil && current.Right == nil {
			decoded.WriteRune(current.Char)
			current = ht.Root
		}
	}

	return decoded.String()
}

func (ht *HuffmanTree) PrintCodes() {
	if len(ht.Codes) == 0 {
		fmt.Println("Нет кодов для отображения")
		return
	}

	fmt.Println("Коды символов Хаффмана:")
	for char, code := range ht.Codes {
		fmt.Printf("%c: %s\n", char, code)
	}
}

func (ht *HuffmanTree) PrintTree() {
	fmt.Println("Дерево Хаффмана:")
	ht.printNode(ht.Root, "")
}

func (ht *HuffmanTree) printNode(node *Node, prefix string) {
	if node == nil {
		return
	}

	if node.Left == nil && node.Right == nil {
		fmt.Printf("%s'%c' (freq: %d)\n", prefix, node.Char, node.Frequency)
		return
	}

	fmt.Printf("%s* (freq: %d)\n", prefix, node.Frequency)

	if node.Left != nil {
		ht.printNode(node.Left, prefix+"  ")
	}
	if node.Right != nil {
		ht.printNode(node.Right, prefix+"  ")
	}
}

func (ht *HuffmanTree) GetCompressionRatio(originalText string) float64 {
	if originalText == "" {
		return 0
	}

	originalBits := len(originalText) * 8
	encodedBits := len(ht.Encode(originalText))

	if encodedBits == 0 {
		return 0
	}

	return float64(originalBits) / float64(encodedBits)
}
